lib/chef/provider/package/windows/exe.rb

```diff
@@ -0,0 +1,159 @@
+            end
+            Chef::Log.info("Removing #{@new_resource} with uninstall command '#{uninstall_command}'")
+            shell_out!(uninstall_command, { returns: @new_resource.returns })
+          end
+
+          private
+
+          def current_installed_version
+            @current_installed_version ||= begin
+              if installed_packages.include?(@new_resource.package_name)
+                installed_packages[@new_resource.package_name][:version]
+              end
+            end
+          end
+
+          def installed_packages
+            @installed_packages || begin
+              installed_packages = {}
+              # Computer\HKEY_LOCAL_MACHINE\Software\Microsoft\Windows\CurrentVersion\Uninstall
+              installed_packages.merge!(extract_installed_packages_from_key(::Win32::Registry::HKEY_LOCAL_MACHINE)) # rescue nil
+              # 64-bit registry view
+              # Computer\HKEY_LOCAL_MACHINE\Software\Wow6464Node\Microsoft\Windows\CurrentVersion\Uninstall
+              installed_packages.merge!(extract_installed_packages_from_key(::Win32::Registry::HKEY_LOCAL_MACHINE, (::Win32::Registry::Constants::KEY_READ | 0x0100))) # rescue nil
+              # 32-bit registry view
+              # Computer\HKEY_LOCAL_MACHINE\Software\Wow6432Node\Microsoft\Windows\CurrentVersion\Uninstall
+              installed_packages.merge!(extract_installed_packages_from_key(::Win32::Registry::HKEY_LOCAL_MACHINE, (::Win32::Registry::Constants::KEY_READ | 0x0200))) # rescue nil
+              # Computer\HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Uninstall
+              installed_packages.merge!(extract_installed_packages_from_key(::Win32::Registry::HKEY_CURRENT_USER)) # rescue nil
+              installed_packages
+            end
+          end
+
+          # http://unattended.sourceforge.net/installers.php
+          def unattended_installation_flags
+            case @installer_type
+            when :installshield
+              '/s /sms'
+            when :nsis
+              '/S /NCRC'
+            when :inno
+              # "/sp- /silent /norestart"
+              '/verysilent /norestart'
+            when :wise
+              '/s'
+            end
+          end
+
+          def extract_installed_packages_from_key(hkey = ::Win32::Registry::HKEY_LOCAL_MACHINE, desired = ::Win32::Registry::Constants::KEY_READ)
+            uninstall_subkey = 'Software\Microsoft\Windows\CurrentVersion\Uninstall'
+            packages = {}
+            begin
+              ::Win32::Registry.open(hkey, uninstall_subkey, desired) do |reg|
+                reg.each_key do |key, _wtime|
+                  begin
+                    k = reg.open(key, desired)
+                    display_name = begin
+                                     k['DisplayName']
+                                   rescue
+                                     nil
+                                   end
```

Comment:
I think we are gonna take a different strategy here. As @btm points out, there can be multiple versions of a "package" installed that share the same display name. We should really be using the leaf key as a unique identifier. So our "current_version" would possibly be a collection of versions of all installed packages with the same display name. We would converge if:
* the new_version is different from all installed version
* there are no installed versions

Further, currently a new_version can only be specified via an attribute. I believe we can improve on this by looking at the file properties (this is different from the properties exposed via an MSI). I know that `inno` and null soft (probably wise and install shield too) expose file properties that line up with the installed version. We can make a "best effort" to grab those and avoid a converge if they match with any installed version.